### ISSUE TITLE
Unable to start server: Failed to load libsteam.so

### ISSUE THREAD

**Issue by kajmaj87** [jammsen/docker-the-forest-dedicated-server#34]
Hi,

I'm unable to start the image on a CentOS 8 server and I'm running out of ideas to check. When starting I get the following output:

```
the-forest-dedicated-server | Checking if Wine is set in bashrc
the-forest-dedicated-server | Setting up Wine in bashrc
the-forest-dedicated-server | Redirecting stderr to '/root/Steam/logs/stderr.txt'
the-forest-dedicated-server | ILocalize::AddFile() failed to load file "public/steambootstrapper_english.txt".
the-forest-dedicated-server | [  0%] Checking for available update...
the-forest-dedicated-server | [  0%] Download Complete.
the-forest-dedicated-server | [----] Applying update...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Installing update...
the-forest-dedicated-server | [----] Installing update...
the-forest-dedicated-server | [----] Failed to apply update, reverting...
the-forest-dedicated-server | [----] Verifying installation...
the-forest-dedicated-server | [  0%] Downloading Update...
the-forest-dedicated-server | [  0%] Checking for available update...
the-forest-dedicated-server | [  0%] Download Complete.
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Extracting package...
the-forest-dedicated-server | [----] Installing update...
the-forest-dedicated-server | [----] Installing update...
the-forest-dedicated-server | [----] Failed to apply update, reverting...
the-forest-dedicated-server | [----] !!! Fatal Error: Failed to load libsteam.so
the-forest-dedicated-server | wine: failed to open "/theforest/TheForestDedicatedServer.exe": c0000135
the-forest-dedicated-server exited with code 53
```

I've tried to run  the image on my laptop running manjaro and also in a clean CentOS 8 Virtual Machine and both started without an issue. But when running this on VPS I always get this issue.

I've been suspecting firewall issues so I've opened ports 8766 and 27000-27100 but this didn't help. This happens even when I opened everything in iptables and disabled firewalld. Any ideas why this might happen? Is this even network related?

**Comment by jammsen** [comment]
Wine tells you the error, it has nothing to do with firewall:
wine: failed to open "/theforest/TheForestDedicatedServer.exe": c0000135

Try totally fresh reinstalling of the container and check for missing file permissions.

**Comment by kajmaj87** [comment]
Thanks for the response. Any ideas what misconfiguration could cause this? The only adjustments that I did was to provide the API token as written in readme. 
For me the config issue seems unlikely as the same docker-compose file that I use (without any config adjustments outside of it) worked on two different linux machines without the above error.
Which file permissions should I check?

I've reinstalled (rm all directories that are mounted, docker-compose down & docker-compose up) at least 10 times during my tries to fix the issue.

**Comment by jammsen** [comment]
Whenever you get something of a .exe isnt found error your install wasnt successful because of something, most of the time missing rights for your user or the docker user. You can see here what happend in the log with the .exe handling:
https://github.com/jammsen/docker-the-forest-dedicated-server/blob/master/usr/bin/servermanager.sh#L78-L84

Also dont forgot the line above that:
the-forest-dedicated-server | [----] !!! **Fatal Error: Failed to load libsteam.so**
the-forest-dedicated-server | wine: failed to open "/theforest/TheForestDedicatedServer.exe": c0000135

Your SteamCMD seems to have a problem, therefore the following after that might not work.The log is verbose on purpose, try googling what might caused that SteamCMD error. 

I dont know nothing about your setup, cant help like this any more except make suggesstions.

**Comment by kajmaj87** [comment]
Yes, I've noticed the highlighed error (I've put it in the issue title). I've google about this, but most issues showing are from around 2013 and concern local setups. 
The dockerimage should be self-contained, right? It's not like it takes any steam related stuff from system. The same setup worked on another CentOS 8 fresh install in VM thats why I suspected network issues, which might have caused erronous setup from SteamCMD inside the container.

> most of the time missing rights for your user or the docker user

Any more suggestions on this, what could be missing for the user? If the user is in docker group and can successfully start containers then all the other permissions in use should be governed by the Dockerfile or am I missing something?

**Comment by jammsen** [comment]
Q: If the user is in docker group and can successfully start containers then all the other permissions in use should be governed by the Dockerfile or am I missing something?
A: In a nutshell this should be it.

Q: Any more suggestions on this, what could be missing for the user?
A: Again i know nothing about your setup, how should i even begin to answer this?

**Comment by kajmaj87** [comment]
What info about my setup would help you to help me? :)

**Comment by jammsen** [comment]
No offense, but i really would appreciate when people read my readme.MD ( https://github.com/jammsen/docker-the-forest-dedicated-server/blob/master/README.md ) - espacially the part about "**What you need to run this**" it says: "**Basic understanding of Linux and Docker**". Its not just there for the lols. It implies an understanding of the tools, _you_ try to run and when you create an issue, the ability to understand your problem, because you invested time into it and help me, help you to find a solution, in other words: Work on this together. In a picture spoken i wont lead a horse to the water, point on it and then force it to drink, it has to be "i help you, to help yourself".

So maybe start by describing your "setup" what OS, what Docker, how is the setup of Docker and all that kind of stuff you use Docker ... again you should know your setups, your should know whats special, you should know whats important, all you need to do is to explain it, after that describe what you did and where you did it to get the server running in a working environment and then on the "not working" system.

**Comment by kajmaj87** [comment]
Before contacting you:
- I've read your readme
- I've read your code, your scripts, your dockerfile
- I've used Linux as my daily OS since around 12 years
- I've used Docker at work daily since at least 2 years (I think those could qualify as me being confident in having **Basic understanding of Linux and Docker**)
- I've tried many ideas, including installing this on my laptop, installing a VM to check this out, fiddling with firewall, connecting into container and trying various things there, installing libraries, retrying installation etc
- I've spend at least 4 hours of research and testing things out before I've run out of ideas and contacting you in hope of that that I've missed something obvious here

Really not sure why are you being aggressive about this and why you assume I'm not willing to work on this or that I did not do my research. Its yours and mine free time, we both are aware of that. If you do not want to help me its perfectly fine with me and I understand this, just please do not attack me as we both have better things to do. Maybe instead just be direct and write that you do not want to help.

And I'm really clueless on what would the knowlage of my docker version and os version (I'm using CentOS 8 on my VPS) give you, but if this would be helpful I'd gladly provide that in the evening.

**Comment by jammsen** [comment]
_Really not sure why are you being aggressive about this and why you assume I'm not willing to work on this or that I did not do my research._
I am not aggressive and i wrote not offense just at the start, so please dont take it the wrong way. I just got the feeling you did not had the basics down.

_Its yours and mine free time, we both are aware of that. If you do not want to help me its perfectly fine with me and I understand this, just please do not attack me as we both have better things to do. Maybe instead just be direct and write that you do not want to help._
Yes to free time, No to "if you dont want to help", if i wouldnt want to help you, i wouldnt pause work for answering you.

_And I'm really clueless on what would the knowlage of my docker version and os version (I'm using CentOS 8 on my VPS) give you, but if this would be helpful I'd gladly provide that in the evening._

You explaining your setup and OS/Arch would help me to google better the error - like for example this https://github.com/ValveSoftware/steam-for-linux/issues/1981 

Thats why i wanted you to explain, is Docker as an example rootful or rootless installed? Which directory you working in? Does it have the right chmod or chown permissions? Do you maybe use a very old Docker Version where you could install a update to fix bugs in Docker itself, that help to resolve this issue?

I mean you dont have to provide any information to me, but it shortens the time where i have to write "no idea bud .... cant help you sorry".

**Comment by kajmaj87** [comment]
> Thats why i wanted you to explain, is Docker an example rootful or rootless installed

I'm using rootless docker, I've created a separate user for running this container and added it to docker group, which I've mentioned above.

> Why directory you working in? Does it have the right chmod or chown permissions?

I'm running this in my users home directory and I've created the directories steamcmd, game and winedata with the rootless user so I assume permission should be fine, after running the container the directories have new files in them.

> You explaining your setup and OS/Arch would help me to google better the error - like for example this https://github.com/ValveSoftware/steam-for-linux/issues/1981 

I've seen this issue (which is from 2013 as other issues linking to libsteam.so, which I've also mentioned above) and tried to install analogue debian libraries inside the container which didn't have any effect on the error. As I also mentioned I've used the same docker-compose on a fresh CentOS8 VM image (same OS as in my VPS where I have issues) also using rootless docker run with a user in docker group and it worked without issues, so installing additional libraries on the container doesn't really have a point.
I'm not sure my OS/Arch is relevant here (arch is standard x86_64 bit arch) - Dockerfile installs all the additional 32bit libraries needed for 64bit OS and should work on any OS with 64bit architecture. That's why I was suspecting something related to networking as this is the only thing that comes to my mind that could affect different behaviour of the same image on same OS but one being on my laptop and the other in some datacenter. Turning off firewall should solve that but didn't and that's when I became clueless.

> Do you maybe use a very old Docker Version where you could install a update to fix bugs in Docker itself, that help to resolve this issue?

I cannot check the exact version now, I would be able to do this first in the evening.

> I mean you dont have to provide any information to me, but it shortens the time where i have to write "no idea bud .... cant help you sorry".

I'm glad to provide any information that would be helpful, it's just the amount of information I could provide is quite big if I'm not sure what would be relevant for you to analyse this.

**Comment by jammsen** [comment]
Hmmm quite a weird error tbh, im not that deep into CentOS and maybe a hidden weirdness between your 2 systems or so. Docker Rootless shouldnt be that big of a problem if you follow good permission schemas and check up on it. I would say its down to either disk been full or permission not fitting.

You could use this article with Docker Rootless to understand what does Docker with which user when you dont have User Remap active or a GID/UID inside the Dockerfile. that article goes very deep and i recommend try things for your Docker Rootless out and compare your results on both systems:
https://medium.com/@mccode/understanding-how-uid-and-gid-work-in-docker-containers-c37a01d01cf

Also ... it might be something totally different, but i have no other idea, with the information at hand.
If you like you could provide the results you tested with that article.

**Comment by jammsen** [comment]
_I'm glad to provide any information that would be helpful, it's just the amount of information I could provide is quite big if I'm not sure what would be relevant for you to analyse this._

In my last 10 years of experience, i found its more helpful to have more information to sort through, that having less and or none. Maybe its just me, but i helps me to think things through and sort things out. might be totally different for you :)

**Comment by kajmaj87** [comment]
Thanks, I'll check the suggestions when I get time in the afternoon and get back to you.

**Comment by jammsen** [comment]
Do that, happy weekend

**Comment by kajmaj87** [comment]
Ok, I've updated the docker version and after that the error message was more descriptive stating that permissions in winedata are wrong.
It turned out that the winedata was owned by my user but the container run as root user (I misunderstood what rootless means, I assumed its just that that the non-root user can use docker commands). Wine refused to write new stuff to directory owned by other user although it would be possible permissionwise.
I fixed this by adding my uid and gid to docker-compose.yaml.

Thanks for all your help and your suggestions.
